src/bio/bio_monitor.c

```diff
@@ -0,0 +1,596 @@
+	D_PRINT("SPDK Device Controller Data [%s]:\n",
+		spdk_bdev_get_name(bdev));
+	D_PRINT("==========================================================\n");
+	D_PRINT("Vendor ID: %04x\n", cdata->vid);
+	D_PRINT("Serial Number: ");
+	dprint_ascii_string(cdata->sn, sizeof(cdata->sn));
+	D_PRINT("\n");
+	D_PRINT("Model Number: ");
+	dprint_ascii_string(cdata->mn, sizeof(cdata->mn));
+	D_PRINT("\n");
+	D_PRINT("Firmware Version: ");
+	dprint_ascii_string(cdata->fr, sizeof(cdata->fr));
+	D_PRINT("\n");
+	D_PRINT("Error log page entries supported: %d\n", cdata->elpe + 1);
+
+prep_cmd:
+	/* Prep NVMe command to get device error log pages */
+	ep_sz = sizeof(struct spdk_nvme_error_information_entry);
+	numd = ep_sz / sizeof(uint32_t) - 1u;
+	numdl = numd & 0xFFFFu;
+	numdu = (numd >> 16) & 0xFFFFu;
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.opc = SPDK_NVME_OPC_GET_LOG_PAGE;
+	cmd.nsid = SPDK_NVME_GLOBAL_NS_TAG;
+	cmd.cdw10 = numdl << 16;
+	cmd.cdw10 |= SPDK_NVME_LOG_ERROR;
+	cmd.cdw11 = numdu;
+	if (cdata->elpe >= DAOS_MAX_ERROR_LOG_PAGES) {
+		D_ERROR("Device error log page size exceeds buffer size\n");
+		dev_health->bhm_inflights--;
+		goto out;
+	}
+	ep_buf_sz = ep_sz * (cdata->elpe + 1);
+
+	/*
+	 * Submit an NVMe Admin command to get device error log page
+	 * to the bdev.
+	 */
+	rc = spdk_bdev_nvme_admin_passthru(dev_health->bhm_desc,
+					   dev_health->bhm_io_channel,
+					   &cmd,
+					   dev_health->bhm_error_buf,
+					   ep_buf_sz,
+					   get_spdk_err_log_page_completion,
+					   dev_health);
+	if (rc) {
+		D_ERROR("NVMe admin passthru (error log), rc:%d\n", rc);
+		dev_health->bhm_inflights--;
+	}
+
+out:
+	/* Free I/O request in the competion callback */
+	spdk_bdev_free_io(bdev_io);
+}
+
+static void
+get_spdk_log_page_completion(struct spdk_bdev_io *bdev_io, bool success,
+			     void *cb_arg)
+{
+	struct bio_health_monitoring			*dev_health = cb_arg;
+	struct spdk_nvme_health_information_page	*hp;
+	struct bio_device_health_state			*hs;
+	struct spdk_bdev				*bdev;
+	struct spdk_nvme_cmd				 cmd;
+	uint32_t					 cp_sz;
+	uint8_t						 crit_warn;
+	int						 rc;
+	int						 sc, sct;
+
+	D_ASSERT(dev_health->bhm_inflights == 1);
+
+	/* Additional NVMe status information */
+	spdk_bdev_io_get_nvme_status(bdev_io, &sct, &sc);
+	if (sc) {
+		D_ERROR("NVMe status code/type: %d/%d\n", sc, sct);
+		dev_health->bhm_inflights--;
+		goto out;
+	}
+
+	D_ASSERT(dev_health->bhm_io_channel != NULL);
+	bdev = spdk_bdev_desc_get_bdev(dev_health->bhm_desc);
+	D_ASSERT(bdev != NULL);
+	hp = dev_health->bhm_health_buf;
+
+	/* Store device health info in in-memory health state log. */
+	hs = &dev_health->bhm_health_state;
+	rc = get_health_timestamp(&hs->bhs_timestamp);
+	if (rc) {
+		D_ERROR("Could not update dev health monitoring timestamp\n");
+		dev_health->bhm_inflights--;
+		goto out;
+	}
+	hs->bhs_temperature = hp->temperature;
+	crit_warn = hp->critical_warning.bits.temperature;
+	hs->bhs_temp_warning = crit_warn;
```

Comment:
why don't just assign hp->critical_warning.bits.temperature to hs->bhs_temp_warning directly? Also, the hp->critical_warning.bits.temperature is just one bit, but the hs->bhs_temp_warning is 1 byte.